Refactor AI::Step's "Find someone to assist me" section

tehhowch opened the pull request:
Ref #2835 , #2846, perhaps some steam issues.

This PR dramatically reduces the game lockup seen when entering an uninhabited system with a large number of out-of-fuel escorts, to the point where it is almost indistinguishable from that of the normal game.

Summary
 - Move "can anyone help me" code into into separate function
 - Use std::vector to store possible choices instead of 2 pointers.
 - Randomly choose a possible helper (must check all ships in the system in order to determine if there is an enemy present)
 - Reduce frequency of calls to avg. 2 per second rather than every step for every stranded/disabled ship.

endless-sky commented:
Would this make it so that twice a second, each ship can potentially change which ship it's asking for help? The advantage of the previous system was that in general, a ship will ask the same ship for help over and over gain.

Also, this would make it so that a ship that's out of fuel will come to a stop as if waiting to be boarded even if it has not succeeded in requesting help from anyone (because 29 frames out of 30, it skips the check and just assumes someone is coming to help it).

tehhowch commented:
I believe those concerns are now addressed:
 - the ship that gets asked is stored in a map, and removed when that ship decides that it cannot help the requestor.
 - a ship with no one helping it will not stand still until it finds someone to help it.

Also, the current implementation on master will never set `selectNext` to true, because at that point in the code `ship` and `it` can not be the same ("Never ask yourself for help"). This bit hadn't been updated since its original addition in https://github.com/endless-sky/endless-sky/commit/d87f009b55b6faeb092a8ed92dc57e26c3f150ab :)
